### DISCUSSION TITLE
How can I change to another tab programmatically?

### DISCUSSION THREAD

**Discussion by maggie44** [quasarframework/quasar#14250] [General - Components / Directives / etc] [answered]
Tabs are working great, and when I click one it slides across and shows a new tab-panel. How can I replicate that same thing though a button or from within my code so that I can change the tab panel without needing to click?

Here is an example in Codepen. When I click a tab, it slides no problem, but when I click the button it is blank:

https://codepen.io/maggie0002/pen/VwXNzae

EDIT: Holding that thought, the codepen one is working, I have added the code in the below comment.

**Comment by pdanpdan** [comment]
Hello. Read the Vue docs on how refs work

**Comment by maggie44** [comment]
Urgh, it's hard to replicate in Codepen, but I realise it's the best way to go about it. In Code pen the tabs don't even seem to want to display horizontally. 

The broad goal is to have the first tab be a 'welcome' tab which is permanent, then the remainder of the tabs generate with v-for so they can be added to from the code. Then each of the tabs other than 'welcome' will have a close button on it to close the tab. I have managed to get all this working, except when I click close and want to return the user back to the welcome tab it just displays a blank page. 

Here is the stripped back code for now, and will have another go at trying to get a codepen going. What's particularly strange about this is the `currentTab.value = reqHost` works but the `currentTab.value = 'welcome'` doesn't:

```
<template>
  <q-tabs
    v-model="currentTab"
    no-caps
    outside-arrows
    mobile-arrows
    class="bg-accent shadow-2"
  >
    <q-tab
      name="welcome"
      :label="$t('components.layouts.captive_portal.welcome')"
    />
    <q-tab
      v-for="(tab, index) in tabs"
      :key="tab"
      :label="tab.title"
      :name="tab.title"
      @click="setAxios(tab.title)"
    >
      <q-badge color="accent" floating transparent>
        <q-icon name="close" color="black" @click="closeTab(index)" />
      </q-badge>
    </q-tab>
  </q-tabs>

  <q-tab-panels v-model="currentTab" animated swipeable>
    <q-tab-panel class="q-pa-lg text-center" name="welcome">
      <q-input
        v-model="reqHostname"
        clearable
        placeholder="balena.local"
        class="q-mt-md"
        filled
        hide-bottom-space
        :label="$t('general.hostname')"
        @keydown.enter.prevent="setHostname(reqHostname)"
      >
        <template #after>
          <q-btn
            v-bind="qBtnStyle"
            flat
            icon="send"
            @click="setHostname(reqHostname)"
          /> </template
      ></q-input>
    </q-tab-panel>

    <q-tab-panel v-for="tab in tabs" :key="tab.title" :name="tab.title">
      <div class="q-pa-lg">
        <device-info />
      </div>
    </q-tab-panel>
  </q-tab-panels>
</template>

<script lang="ts">
import DeviceInfo from 'components/SystemDeviceInfo.vue'
import { qHeaderStyle } from 'src/config/qStyles'
import { axiosUrl } from 'stores/system'
import { defineComponent, ref } from 'vue'

interface tabIndex {
  title?: string
}

export default defineComponent({
  name: 'LayoutsCaptivePortal',
  components: {
    DeviceInfo
  },
  setup() {
    const axiosBaseUrl = axiosUrl()
    const currentTab = ref('welcome')
    const reqHostname = ref('')
    const tabs = ref<tabIndex[]>([])

    function closeTab(index: number) {
      tabs.value.splice(index, 1)
      currentTab.value = 'welcome'
    }

    function setAxios(currentTabSetting: string) {
      axiosBaseUrl.setUrl(`http://${currentTabSetting}`)
    }

    function setHostname(reqHost: string) {
      tabs.value?.push({
        title: reqHost
      })
      setAxios(reqHost)
      currentTab.value = reqHost
      reqHostname.value = ''
    }

    return {
      closeTab,
      currentTab,
      qHeaderStyle,
      reqHostname,
      setAxios,
      setHostname,
      tabs
    }
  }
})
</script>
```

**Reply by metalsadman** [reply]
works fine here https://codepen.io/metalsadman/pen/wvvjepy it's in vue2/quasar v1 but should work the same i recon.

here's quasar v2 version (it's not in composition api)
https://codepen.io/metalsadman/pen/ExEJbPQ

**Reply by maggie44** [reply]
Incredibly helpful, thanks! I'm trying to work backwards from that to find a fix. I have stripped it back and mimicked a lot, but still seems to be doing the same issue. 

A potential clue though although I can't figure out what it means myself. If I insert `{{currentTab}}` somewhere on the page so I can see what is stored in that v-model, when I click the close button it doesn't change. It changes when I click a tab manually, displaying the item name of the correct page. But when using the close button which manipulates that var from the script it doesn't change. This would explain why it is displaying blank, but not why it doesn't change.

So I'm thinking it must have something to do with the reactivity of the `currentTab` variable passed to the tabs v-model? When changing it in the `<script lang="ts">` it isn't reflected in the page. Strangely it is in the function which calls `currentTab.value = reqHost` but not the one that calls `currentTab.value = 'welcome'`.

I have tried changing it with the click `<q-icon name="close" color="black" @click="currentTab = 'welcome'" />` but it flashes to the welcome page then comes back.

**Comment by maggie44** [comment, marked as answer]
Sometimes it is the simplest things, wrapped in complexity. It seems clicking the close button was also sending a click request through to the underlying tab, which was then telling it to come back again. Changed my click event to `@click.stop="closeTab(index)"` solved the issue.

Massive thanks for all the help on this, it consumed so much time but glad to have a solution, and a bunch of ideas from the code example provided that has improved what I had too.
